compiler/rustc_lint/src/context.rs

```diff
@@ -419,29 +419,49 @@ impl LintStore {
         }
     }
 
-    fn no_lint_suggestion(&self, lint_name: &str) -> CheckLintNameResult<'_> {
+    fn no_lint_suggestion(&self, lint_name: &str, tool_name: &str) -> CheckLintNameResult<'_> {
         let name_lower = lint_name.to_lowercase();
 
         if lint_name.chars().any(char::is_uppercase) && self.find_lints(&name_lower).is_ok() {
             // First check if the lint name is (partly) in upper case instead of lower case...
-            return CheckLintNameResult::NoLint(Some(Symbol::intern(&name_lower)));
+            return CheckLintNameResult::NoLint(Some((Symbol::intern(&name_lower), false)));
         }
+
         // ...if not, search for lints with a similar name
-        // Note: find_best_match_for_name depends on the sort order of its input vector.
-        // To ensure deterministic output, sort elements of the lint_groups hash map.
         // Also, never suggest deprecated lint groups.
-        // We will soon sort, so the initial order does not matter.
         #[allow(rustc::potential_query_instability)]
-        let mut groups: Vec<_> = self
+        let groups: Vec<_> = self
             .lint_groups
             .iter()
             .filter_map(|(k, LintGroup { depr, .. })| depr.is_none().then_some(k))
             .collect();
-        groups.sort();
         let groups = groups.iter().map(|k| Symbol::intern(k));
         let lints = self.lints.iter().map(|l| Symbol::intern(&l.name_lower()));
-        let names: Vec<Symbol> = groups.chain(lints).collect();
-        let suggestion = find_best_match_for_name(&names, Symbol::intern(&name_lower), None);
+        let mut names: Vec<Symbol> = groups.chain(lints).collect();
+        // Here, a little bit of extra hack
+        // we use the lint names from rustc and to mock the tool name,
+        // if it's selected we then strip to the right suggestion with `true` for from_rustc
+        // so we can handle these cases:
+        //  1. suggest `missing_docs` from rustc for `clippy::missing_docs`
+        //  2. suggest `dead_code` from rustc for `clippy::dead_cod`
+        let rustc_names = self
+            .by_name
+            .keys()
+            .map(|k| Symbol::intern(&format!("{tool_name}::{k}")))
+            .collect::<Vec<_>>();
+        names.extend(rustc_names.clone());
```

Comment:
There's a bit too much cloning and allocation of potentially big vectors going on here. I would slightly refactor this and first just check the `names` Vec and if a similar name is found, early return. Only if the lint couldn't be found in the `names` I would try to strip the `name_lower` from the `tool_name` and just retry the `find_best_match_for_name` on the stripped `name_lower` with the `names` Vec. That way, you don't even have to build `rustc_names`. This should lead to the same result, unless I'm missing an edge case here.